src/mlir/design.md

```diff
@@ -0,0 +1,158 @@
+# Verona MLIR Generator Design Choices
+
+The MLIR generator sits in between the AST and LLVM IR, and is meant to ease
+lowering while still maintaining some high level concepts to allow us to optimise
+the code without having to scan the whole IR on every pass.
+
+The LLVM pipeline intermixes analysis passes with transformation passes and has
+a way of tracking which transformation pass invalidates which analysis pass to
+avoid re-running complex analysis passes. However, more often than not, the
+transformation passes end up invalidating them anyway and a new run is needed.
+
+In Verona, we'll be bundling all modules, including the standard library and all
+source files, into one module. A full scan on that module for every analysis pass
+would be prohibitively expensive.
+
+LLVM IR has a fixed shape, with instructions belonging to basic blocks, which in
+turn belong to functions. Instructions have "uses", which we can be traversed
+cheaply, attributes and metadata, which can belong to more than one symbol, but
+that only gets us so far.
+
+Verona needs to know the type constraints of objects (imm, mut, iso) as well as
+which region they belong to and if those regions have the same sentinel object.
+We may also want to do intra-module optimisations if we know a symbol doesn't
+escape the context (which is much cheaper than whole program optimisation).
+
+For those reasons, Verona will need a light dialect to represent type constraints,
+region information and module boundaries, which don't naturally convert to LLVM.
+
+## Generation Simplicity
+
+The new AST is much simpler than the previous. The new parser uses a lean typed
+AST and converts all language into those nodes, which becomes much simpler to
+lower to an IR. The parser also runs the type checks and inference, so the AST
+we get at this stage is already complete, and checked for syntactic and semantic
+errors.
+
+Lowering errors can still occur, if the semantics would be correct in some cases
+but not in this specific case, for example, when using it with a different
```

Comment:
It would be good to have an example of this, I don't know what it means.

Reply:
I have no idea either, but I'm leaving it open to errors that can only be caught in MLIR. I'll rewrite it to be more vague.